yellowstone-grpc-tools/src/scylladb/agent.rs

```diff
@@ -0,0 +1,331 @@
+        Ok(receiver)
+    }
+
+    pub fn kill(self) {
+        self.abort_handle.abort();
+    }
+}
+
+pub struct AgentSystem {
+    pub default_agent_buffer_capacity: usize,
+    handlers: JoinSet<anyhow::Result<Nothing>>,
+}
+
+impl AgentSystem {
+    pub fn new(default_agent_buffer_capacity: usize) -> Self {
+        AgentSystem {
+            default_agent_buffer_capacity,
+            handlers: JoinSet::new(),
+        }
+    }
+
+    pub fn spawn<T, N: Into<String>>(&mut self, name: N, ticker: T) -> AgentHandler<T::Input>
+    where
+        T: Ticker + Send + 'static,
+    {
+        self.spawn_with_capacity(name, ticker, self.default_agent_buffer_capacity)
+    }
+
+    pub fn spawn_with_capacity<T, N: Into<String>>(
+        &mut self,
+        name: N,
+        mut ticker: T,
+        buffer: usize,
+    ) -> AgentHandler<T::Input>
+    where
+        T: Ticker + Send + 'static,
+    {
+        let (sender, mut receiver) = channel::<Message<T::Input>>(buffer);
+        let agent_name: String = name.into();
+        let inner_agent_name = agent_name.clone();
+        let abort_handle = self.handlers.spawn(async move {
+            let name = inner_agent_name;
+
+            let init_result = ticker.init().await;
+
+            if init_result.is_err() {
+                error!("{:?} error during init: {:?}", name, init_result);
+                return init_result;
+            }
+
+            loop {
+                let before = Instant::now();
+                let result = tokio::select! {
+                    _ = ticker.timeout() => {
+                        ticker.on_timeout(Instant::now()).await
+                    }
+                    opt_msg = receiver.recv(), if ticker.is_pull_ready() => {
+                        match opt_msg {
+                            Some(msg) => {
+                                let now = Instant::now();
+                                if msg.callbacks.is_empty() {
+                                    ticker.tick(now, msg.data).await
+                                } else {
+                                    ticker.tick_with_callback_sender(now, msg.data, msg.callbacks).await
+                                }
+                            },
+                            None => {
+                                let now = Instant::now();
+                                return Err(
+                                    ticker.terminate(now)
+                                        .await
+                                        .err()
+                                        .unwrap_or(anyhow!("Agent is closed"))
+                                )
+                            },
+                        }
+                    }
+                };
+
+                if result.is_err() {
+                    let emsg = format!("message loop: {:?}, {:?}", name, result.err().unwrap());
+                    error!(emsg);
+                    return Err(anyhow!(emsg));
+                }
```

Comment:
```suggestion
                if let Err(error) = result {
                    let emsg = format!("message loop: {name:?}, {error:?}");
                    error!(emsg);
                    anyhow::bail!(emsg);
                }
```
should we log if return error? :thinking:

Reply:
Nice I was not aware of `bail!` macro.